Another New Dispatcher

ajohnstonTE opened the pull request:
WIP

ajohnstonTE committed:
Add gemini-jax-rs module and explanation to self on how to handle path dispatching

ajohnstonTE committed:
WIP implement dispatcher

ajohnstonTE committed:
Add support for regex variables

ajohnstonTE committed:
Add semi-temporary performance tests

A lot of this will be removed eventually. Committing it just makes it easier to distinguish
from any new functionality and some near-future refactors while doing dev.

msmith-techempower commented on gemini-jax-rs/src/main/java/com/techempower/gemini/jaxrs/core/JaxRsDispatcher.java:
```diff
@@ -0,0 +1,911 @@
+    Object instance = matchInfo.endpoint.resource.getInstance();
+    Class<?> instanceClass = matchInfo.endpoint.resource.getInstanceClass();
+    Class<?>[] parameterTypes = method.getParameterTypes();
+    Annotation[][] parameterAnnotations = method.getParameterAnnotations();
```
It seems to me that this reflection work should be done and cached at startup, rather than on a per-request dispatch.

ajohnstonTE replied:
There's a TODO in here somewhere for exactly that. Atm I'm just focusing on matching the spec so I have a better idea of how to structure a lot of this. It entirely possible I'll get to that caching earlier than I expect, but I'm planning on doing that once the structure is less likely to be heavily refactored (which is definitely going to happen before this dispatcher is done).

ajohnstonTE committed:
Update performance tests to avoid bias

ajohnstonTE committed:
Implement media type parsing, plus tests

ajohnstonTE committed:
Reformat code

ajohnstonTE committed:
Refactor MediaTypeData/MediaTypeDataGroup to capture all parameters

Also MediaTypeData now works much more like MediaType, except lazily in one implementation.

ajohnstonTE committed:
Add a functional, non-performant endpoint registry, plus incomplete (but passing) tests

This should help with designing more complicated tests and work as a reference implementation for the performant version. There's also a dispatching test that is added because I was feeling optimistic, but really it's too high level. The endpoint registry test does a better job of testing the target functionality more precisely. I'll likely remove the dispatching test at some point.
